src/controller/handler/agent.go

```diff
@@ -23,33 +72,111 @@ func ConnectAgent(w http.ResponseWriter, r *http.Request) {
     }
     s := time.Now()
     defer func() {
-        WebsocketDisconnect(c);
+        WebsocketDisconnect(c)
         d := time.Now().Sub(s)
         log.Infof("WEBSOCKET %s(lasted=%s)", "closed", d.String())
     }()
     log.Infof("WEBSOCKET %s", "connected")
 
+    // Moon-Authentication: key,uuid; auth[0]=key, auth[1]=uuid
+    auth := strings.Split(r.Header.Get("Moon-Authentication"), ",")
+    if len(auth) != 2 {
+        log.Errorf("Invalid authentication")
+        w.WriteHeader(http.StatusUnauthorized)
+        return
+    }
+
+    // Step 1: verify user
+    u := model.GetUserByAgentkey(auth[0])
+    if u == nil {
+        log.Errorf("Unauthorized key %s", auth[0])
+        w.WriteHeader(http.StatusUnauthorized)
+        return
+    }
+
+    // Step 2: update node info
+    req := new(Request)
+    req.Action = ActionNodeInfo
+    res, err := WebsocketSendRecv(c, req)
+    if err != nil {
+        w.WriteHeader(http.StatusInternalServerError)
+        return
+    }
+
+    if res.Status != StatusOK {
+        log.Errorf("fail to get node-info from agent")
+        w.WriteHeader(http.StatusUnauthorized)
+        return
+    }
+
+    nodeinfo := new(NodeInfo)
+    err = json.Unmarshal(res.Content, nodeinfo)
+    if err != nil {
+        log.Errorf("fail to parse node-info from agent: %s", err.Error())
+        w.WriteHeader(http.StatusUnauthorized)
+        return
+    }
+
+    if n := model.GetNodeByUuid(auth[1]); n == nil {
+        n := new(model.Node)
+        n.Uuid = auth[1]
+        n.Name = nodeinfo.Hostname
+        n.Description = u.Name + "'s host: " + nodeinfo.Hostname
+        n.OwnerId = u.Id
+        n.Nics = nodeinfo.Nics
+        n.Save()
+    } else {
+        n.Name = nodeinfo.Hostname
+        n.Nics = nodeinfo.Nics
+        n.Update()
+    }
+
+    // Step 3: register wsConns
+    ac := &AgentConnection{Ws: c, Send: make(chan []byte), Done: make(chan string)}
+	wsConns[r.Header.Get("Moon-Authentication")] = ac
+
+    // Step 4: wait and exec deploy commands
     for {
-        mT, m, err := c.ReadMessage()
+        req.Action = ActionExecShell
+        req.Content = <- ac.Send
+
+        res, err = WebsocketSendRecv(c, req)
         if err != nil {
-            log.Errorf("WEBSOCKET READ %s", err.Error())
+            w.WriteHeader(http.StatusInternalServerError)
```

Comment:
maybe not internal server error. any broken connection at this moment can raise this error. but okay for now, merely a little weird.